Payment Methods: add PayPal to change payment method form

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

This adds a PayPal button to the list of options displayed by `ChangePaymentMethod`. Selecting this option will create a new PayPal billing agreement for the subscription in question.

![Screen Shot 2021-01-07 at 3 12 05 PM](https://user-images.githubusercontent.com/9310939/103945492-e3deaf00-50fa-11eb-9ee2-6b882ca876e0.png)

Fixes https://github.com/Automattic/wp-calypso/issues/47518

#### Testing instructions

- Apply D54805-code 
- Visit `/purchases` for a site which has at least one subscription.
- Click on the subscription.
- Click "Add payment method" or "Change payment method"
- Verify that "PayPal" is one of the radio button options. Click to select PayPal.
- Click the submit button at the bottom of the form.
- Verify that you are redirected to PayPal to create a new billing agreement.
- Refuse the billing agreement and verify that you are redirected back to the form to select a payment method.
- Click to select PayPal again.
- Confirm the PayPal billing agreement this time and verify that you are redirected back to the subscription page.
- Verify that you see a notification that says your payment method has been assigned.
- Verify that the payment method assigned to the subscription is PayPal (you should see it in the subscription's card).

sirbrillig committed:
Add PayPal payment method object to useAssignablePaymentMethods

sirbrillig committed:
Add placeholder paypal payment processor

sirbrillig committed:
Remove empty items/total passed to CheckoutProvider

These are now optional props.

sirbrillig committed:
Add validation for PaymentProcessorResponse

This way if a payment processor function returns an invalid response,
you will get a better error than "object does not have a type property".

sirbrillig committed:
Add InvalidPaymentProcessorResponseError

This will allow a more standardized way of handling invalid payment
processor responses if the consumer desires.

sirbrillig committed:
Remove requirement for a payment processor SUCCESS payload

It is unknown and the reducer allows it to be optional.

sirbrillig commented:
This just needs to be hooked up to an endpoint. See the `TODO` in the code. 

We'll also need to test the behavior when PayPal is already set for a subscription. The plan is that the PayPal radio button should never appear pre-selected; selecting it means "add a new PayPal subscription". I believe it will work that way currently but it's worth verifying.

sirbrillig committed:
Add labelText option to useCreatePayPal

sirbrillig committed:
Add alternate labelText to PayPal ChangePaymentMethod button

sirbrillig committed:
Make sure useCreatePayPal args are optional

DavidRothstein committed:
Add rudimentary integration with the API endpoint.

DavidRothstein commented:
I just pushed up a new commit that adds very basic integration with the API endpoint from D54805-code.  The integration needs some work, but this is now enough to start doing basic testing.

sirbrillig commented on client/me/purchases/manage-purchase/change-payment-method/index.jsx:
```diff
@@ -284,8 +286,13 @@ async function assignExistingCardProcessor( purchase, { storedDetailsId } ) {
-async function assignPayPalProcessor() {
-	// TODO: create a new paypal billing agreement
+async function assignPayPalProcessor( purchase ) {
+	return wpcomCreatePayPalAgreement( purchase.id, window.location.href, window.location.href ).then(
```
Awesome! 👏  FWIW you can change this promise to return `makeRedirectResponse( data )` and it will do the redirect for you.

DavidRothstein replied:
Ah, thank you!  I threw these changes up in a hurry, but `makeRedirectResponse` looks like exactly what I should have been looking for.  I'm fixing that now.

DavidRothstein committed:
Use makeRedirectResponse() to perform the redirect.

DavidRothstein commented on client/me/purchases/manage-purchase/change-payment-method/index.jsx:
```diff
@@ -387,6 +393,10 @@ function useAssignablePaymentMethods() {
 	} );
 
+	const payPalMethod = useCreatePayPal( {
+		labelText: translate( 'Add a new PayPal account' ),
```
I wonder if "Add a PayPal account" (from https://github.com/Automattic/wp-calypso/issues/47518), or even just "PayPal", would be better wording here for most situations, with "Add a new PayPal account" only used when PayPal is already the current payment method for the subscription (in combination with https://github.com/Automattic/wp-calypso/pull/48583).

It's not a huge difference, but "new" could be a bit confusing if the user already is using PayPal for other subscriptions and plans to use the same PayPal account here, for example.

michaeldcain replied:
Let's do "PayPal", to keep it similar to Checkout and to match the verb-less credit card label:

<img width="500" alt="Screen Shot 2021-01-05 at 4 29 43 PM" src="https://user-images.githubusercontent.com/942359/103701052-5c1a6880-4f73-11eb-9153-8acf515c186f.png">

sirbrillig replied:
I'm ok with either, but I think it's worth saying that I had intentionally changed this away from "PayPal" after a previous discussion we had because this button works differently than the other buttons. If you already have PayPal selected, this button will not be checked and will still create a new billing agreement. I don't feel very strongly either way though so I'll change it as suggested here.

DavidRothstein replied:
>I'm ok with either, but I think it's worth saying that I had intentionally changed this away from "PayPal" after a previous discussion we had because this button works differently than the other buttons. If you already have PayPal selected, this button will not be checked and will still create a new billing agreement.

Definitely agree with that part -- that's why I was saying above that it **should** still be "Add a new PayPal account" (or I guess "New PayPal account" if we're doing away with the verbs) in that particular scenario.

It should hopefully be easy to distinguish between the two cases via something like `getCurrentPaymentMethodId( purchase.payment ) === 'paypal'` (as in https://github.com/Automattic/wp-calypso/pull/48583).

DavidRothstein reviewed (commented):
Note that the screenshot shows the wrong terms of service text for the PayPal option (and there's also a TODO in the existing code about that, "TODO: Make sure we use the correct ToS text for paypal") so just a quick note here that we'll eventually have to fix that as part of this pull request.

sirbrillig committed:
Change paypal success url to purchase page

sirbrillig commented:
Rebased and changed the success URL to the manage purchase page because otherwise the UX looks a little confusing since the radio button that's checked after success will be the first one in the list and that's probably not paypal.

sirbrillig committed:
Revert purchaseUrl as success callback

sirbrillig committed:
Add success/error event handlers for redirect payment methods

sirbrillig committed:
Change error handler to use callback like success handler

sirbrillig committed:
Remove unnecessary destructure of purchase variable

sirbrillig committed:
Make sure paypal box is never checked on load

sirbrillig committed:
Make error message persistent, just in case we redirect

sirbrillig committed:
Change PayPal radio button label if paypal is selected

sirbrillig commented:
I pushed up some changes to handle success and error messages from the most recent version of the endpoints. I also adjusted the PayPal button title as suggested; it's now `PayPal` if PayPal is not used by this subscription and `New PayPal account` if PayPal _is_ used for this subscription.

The success redirect can be a little slow because it has to wait for the purchase, stripe, and all the credit cards to load. We don't really need most of that, but we do need the purchase ID, and right now we treat all loading data the same. If it's an issue we can move the success/error handlers to happen just once the purchase has loaded.

DavidRothstein commented on client/me/purchases/manage-purchase/change-payment-method/index.jsx:
```diff
@@ -446,6 +466,11 @@ function useAssignablePaymentMethods() {
 
+	const payPalMethod = useCreatePayPal( {
+		labelText:
+			currentPaymentMethodId === 'paypal-existing' ? translate( 'New PayPal account' ) : 'PayPal',
```
Should we translate "PayPal" here too?  I'm unclear if it's actually supposed to be translated (since it's a proper name) but it looks like some other places translate it already, and if nothing else doing so gives translators maximum freedom.

sirbrillig replied:
I intentionally didn't, but I don't mind doing so.

nbloomf commented:
The testing instructions worked as described. Nice!

Here's what the change payment method screen looks like when a paypal account is currently assigned:

![Screen Shot 2021-01-07 at 3 12 05 PM](https://user-images.githubusercontent.com/9310939/103945492-e3deaf00-50fa-11eb-9ee2-6b882ca876e0.png)

sirbrillig committed:
Move useCreateAssignablePaymentMethods to own file

sirbrillig committed:
Move processor functions to own file

sirbrillig committed:
Fix PaymentDetails type in calypso-stripe

It can contain sub-objects.

sirbrillig committed:
Move TosText to own file

sirbrillig committed:
Make TOS more generic so it covers paypal

sirbrillig committed:
Always translate "PayPal"

a8ci18n commented on client/me/purchases/manage-purchase/change-payment-method/tos-text.tsx:
```diff
@@ -0,0 +1,39 @@
+	return (
+		<>
+			{ translate(
+				'You agree to our {{tosLink}}Terms of Service{{/tosLink}}, and if you use it to pay for a subscription or plan, you authorize your payment method to be charged on a recurring basis until you cancel, which you can do at any time. You understand {{autoRenewalSupportPage}}how your subscription works{{/autoRenewalSupportPage}} and {{faqCancellingSupportPage}}how to cancel{{/faqCancellingSupportPa ...
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **22** times:
`translate( 'By saving a credit card, you agree to our {{tosLink}}Terms of Service{{/tosLink}}, and if you use it to pay for a subscription or plan, you authorize your credit card to be charged on a recurring basis until you cancel, which you can do at any time. You understand {{autoRenewalSupportPage}}how your subscription works{{/autoRenewalSupportPage}} and {{managePurchasesSupportPage}}how to cancel{{/managePurchasesSupportPage}}.' )` *ES Score: 9*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/paypal-to-change-card-form)

sirbrillig replied:
Unfortunately, this rewrite is necessary to replace "credit card" with "payment method".

sirbrillig commented:
I pushed a few more changes to clean up the code and to update the TOS to read "payment method" instead of "credit card".

DavidRothstein commented on client/me/purchases/manage-purchase/change-payment-method/tos-text.tsx:
```diff
@@ -0,0 +1,39 @@
+	return (
+		<>
+			{ translate(
+				'You agree to our {{tosLink}}Terms of Service{{/tosLink}}, and if you use it to pay for a subscription or plan, you authorize your payment method to be charged on a recurring basis until you cancel, which you can do at any time. You understand {{autoRenewalSupportPage}}how your subscription works{{/autoRenewalSupportPage}} and {{faqCancellingSupportPage}}how to cancel{{/faqCancellingSupportPa ...
```
I think we should add "By saving a payment method," at the beginning of the first sentence (to match the "By saving a credit card," that was there before).  Otherwise the first "it" in the sentence is pretty ambiguous.

sirbrillig replied:
Good point. I mashed together the previous TOS here with the TOS from checkout and I didn't notice that wording issue.

sirbrillig committed:
Add "By saving a payment method" to TOS

a8ci18n commented on client/me/purchases/manage-purchase/change-payment-method/tos-text.tsx:
```diff
@@ -0,0 +1,39 @@
+	return (
+		<>
+			{ translate(
+				'By saving a payment method, you agree to our {{tosLink}}Terms of Service{{/tosLink}}, and if you use it to pay for a subscription or plan, you authorize your payment method to be charged on a recurring basis until you cancel, which you can do at any time. You understand {{autoRenewalSupportPage}}how your subscription works{{/autoRenewalSupportPage}} and {{faqCancellingSupportPage}}how to can ...
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **22** times:
`translate( 'By saving a credit card, you agree to our {{tosLink}}Terms of Service{{/tosLink}}, and if you use it to pay for a subscription or plan, you authorize your credit card to be charged on a recurring basis until you cancel, which you can do at any time. You understand {{autoRenewalSupportPage}}how your subscription works{{/autoRenewalSupportPage}} and {{managePurchasesSupportPage}}how to cancel{{/managePurchasesSupportPage}}.' )` *ES Score: 8*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/paypal-to-change-card-form)

sirbrillig committed:
Switch TOS text to match checkout

sirbrillig commented:
I went ahead and changed the TOS text to match the one in checkout to avoid having to re-translate it for the time being. We can make another PR in the future if we want to update the text to mention "if you use [the payment method] to pay for a subscription".

DavidRothstein committed:
Move "success" handling entirely to the client side, so it is more self-contained and doesn't rely on the behavior of the server

sirbrillig commented:
Failing tests have to do with Gutenberg and are unrelated.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5429803

Thank you @sirbrillig for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5429803).
